README.md
# AWS_S3_client

Client library to interact with the AWS S3 API (and other compatible S3 services, such as Yandex Cloud Storage)

## Usage

A simple usage example:

```
import 'package:aws_s3_client/aws_s3.dart';

main() async {
  Spaces spaces = Spaces(
    region: "region",
    accessKey: "accessKey",
    secretKey: "secretKey",
  );
  for (String name in await spaces.listAllBuckets()) {
    print('bucket: ${name}');
    if (name == 'yourBucket') {
      Bucket bucket = spaces.bucket(name);
      await for (BucketContent content
          in bucket.listContent(prefix: 'test')) {
        print('key: ${content.key}; size: ${content.size}');
      }
    }
  }
  Bucket bucket = spaces.bucket('yourBucket');

  String etag = await bucket.uploadFile(
      'test/test.md', 'README.md', 'text/plain', Permissions.public);
  print('upload: $etag');

  print('done');
}
```

For more usage examples please see tests

## References

* https://docs.aws.amazon.com/general/latest/gr/Welcome.html
* https://cloud.yandex.ru/docs/storage/s3/api-ref/

Thanks https://github.com/nbspou for inspiration of creating this package


CHANGELOG.md
## 0.3.0

- Migrate to null-safety.

## 0.2.0

- isTruncated parameter is initialized
- added session support for upload
- updated dependencies to latest versions

## 0.1.1

- Remove all warnings of pub.dev

## 0.1.0

- Initial version

test/bucket_test.dart
import 'dart:async';
@TestOn('vm')
import 'dart:io';

import 'package:aws_s3_client/aws_s3_client.dart';
import 'package:mime/mime.dart';
import 'package:path/path.dart' as pathInfo;
import 'package:test/test.dart';

import './test_config.dart';

void main() {
  final spaces = Spaces(
      region: TestConfig.region,
      accessKey: TestConfig.accessKey,
      secretKey: TestConfig.secretKey,
      provider: Provider.yandex);
  final bucket = spaces.bucket(TestConfig.bucket);

  setUpAll(() async {});

  test('List buckets', () async {
    final buckets = await spaces.listAllBuckets();
    expect(buckets, contains(TestConfig.bucket));
  });

  test('List content', () async {
    final file = File('test/squirrel.jpeg');
    final file2 = File('test/earth.jpeg');
    final keys = ['squirrel.jpeg', 'earth.jpeg'];
    await bucket.uploadFile(pathInfo.basename(file.path),
        file.readAsBytesSync(), lookupMimeType(file.path)!, Permissions.public);
    await bucket.uploadFile(
        pathInfo.basename(file2.path),
        file2.readAsBytesSync(),
        lookupMimeType(file2.path)!,
        Permissions.public);

    var content = await bucket.listContent().toList();
    expect(content.length, 2);
    content = await bucket.listContent(prefix: 'squirrel').toList();
    expect(content.length, 1);
    content = await bucket.listContent(prefix: 'earth').toList();
    expect(content.length, 1);

    final completer = Completer();
    bucket
        .listContent(
          maxKeys: 1,
        )
        .listen((data) => expect(keys, contains(data.key)),
            onDone: () => completer.complete());
    await completer.future;
  });

  test('upload file to bucket', () async {
    final file = File('test/squirrel.jpeg');
    final eTag = await bucket.uploadFile(pathInfo.basename(file.path),
        file.readAsBytesSync(), lookupMimeType(file.path)!, Permissions.public);
    expect(eTag, isNotEmpty);

    final content = await bucket.listContent(prefix: 'squirrel').toList();
    expect(content.first.key, pathInfo.basename(file.path));
    expect(content.first.eTag, eTag);
    expect(content.first.size, file.lengthSync());
  });

  test('Delete file', () async {
    final file = File('test/squirrel.jpeg');
    await bucket.uploadFile(pathInfo.basename(file.path),
        file.readAsBytesSync(), lookupMimeType(file.path)!, Permissions.public);

    var content = await bucket.listContent(prefix: 'squirrel').toList();
    expect(content.length, 1);
    await bucket.delete(pathInfo.basename(file.path));
    content = await bucket.listContent(prefix: 'squirrel').toList();
    expect(content.length, 0);
  });

  test('Upload file as stream to bucket', () async {
    final file = File('test/earth.jpeg');
    final eTag = await bucket.uploadFileStream(
        pathInfo.basename(file.path),
        file.openRead(),
        file.lengthSync(),
        lookupMimeType(file.path)!,
        Permissions.public);
    expect(eTag, isNotEmpty);

    final content = await bucket.listContent(prefix: 'earth').toList();
    expect(content.first.key, pathInfo.basename(file.path));
    expect(content.first.eTag, eTag);
    expect(content.first.size, file.lengthSync());

    await bucket.delete(pathInfo.basename(file.path));
    expect(await bucket.listContent(prefix: 'earth').toList(), []);
  });
}


test/test_config.dart
class TestConfig {
  static const region = "ru-central1";
  // ATTENTION! Do not store pruduction keys in public repositories!
  static const accessKey = "accessKey";
  static const secretKey = "secretKey";
  static const bucket = "s3-client-test-bucket";
}


lib/src/client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:meta/meta.dart';
import 'package:crypto/crypto.dart';
import 'package:http_client/console.dart' as http;
import 'package:xml/xml.dart' as xml;

import 'bucket.dart';

const defaultChunkSize = 65536;

class ClientException implements Exception {
  final int statusCode;
  final String reasonPhrase;
  final Map<String, String> responseHeaders;
  final String responseBody;
  const ClientException(this.statusCode, this.reasonPhrase,
      this.responseHeaders, this.responseBody);
  String toString() {
    return "DOException { statusCode: ${statusCode}, reasonPhrase: \"${reasonPhrase}\", responseBody: \"${responseBody}\" }";
  }
}

class Client {
  final String region;
  final String accessKey;
  final String secretKey;
  final String? sessionToken;
  final String service;

  @protected
  final http.Client httpClient;

  Client(
      {required this.region,
      required this.accessKey,
      required this.secretKey,
      required this.service,
      this.sessionToken,
      http.Client? httpClient})
      : this.httpClient =
            httpClient == null ? http.ConsoleClient() : httpClient {}

  @protected
  Future<xml.XmlDocument> getUri(Uri uri) async {
    http.Request request = http.Request('GET', uri, headers: http.Headers());
    signRequest(request, contentSha256: await sha256.convert(''.codeUnits));
    http.Response response = await httpClient.send(request);
    BytesBuilder builder = BytesBuilder(copy: false);
    await response.body.forEach(builder.add);
    String body = utf8.decode(builder.toBytes());
    if (response.statusCode != 200) {
      throw ClientException(response.statusCode, response.reasonPhrase,
          response.headers.toSimpleMap(), body);
    }
    xml.XmlDocument doc = xml.XmlDocument.parse(body);
    return doc;
  }

  String _uriEncode(String str) {
    return Uri.encodeQueryComponent(str).replaceAll('+', '%20');
  }

  String _trimAll(String str) {
    String res = str.trim();
    int len;
    do {
      len = res.length;
      res = res.replaceAll('  ', ' ');
    } while (res.length != len);
    return res;
  }

  @protected
  void signRequest(http.Request request,
      {Digest? contentSha256, int expires = 86400}) {
    // Build canonical request
    String httpMethod = request.method;
    String canonicalURI = request.uri.path;
    String host = request.uri.host;
    // String service = 's3';

    DateTime date = DateTime.now().toUtc();
    String dateIso8601 = date.toIso8601String();
    dateIso8601 = dateIso8601
            .substring(0, dateIso8601.indexOf('.'))
            .replaceAll(':', '')
            .replaceAll('-', '') +
        'Z';
    String dateYYYYMMDD = date.year.toString().padLeft(4, '0') +
        date.month.toString().padLeft(2, '0') +
        date.day.toString().padLeft(2, '0');

    /*dateIso8601 = "20130524T000000Z";
    dateYYYYMMDD = "20130524";
    hashedPayload = null;*/
    String hashedPayloadStr =
        contentSha256 == null ? 'UNSIGNED-PAYLOAD' : '$contentSha256';

    String credential =
        '${accessKey}/${dateYYYYMMDD}/${region}/${service}/aws4_request';
    // Build canonical headers string
    Map<String, List<String>> headers = Map<String, List<String>>();
    request.headers.add('x-amz-date', dateIso8601); // Set date in header
    if (contentSha256 != null) {
      request.headers.add('x-amz-content-sha256',
          hashedPayloadStr); // Set payload hash in header
    }
    request.headers.keys.forEach((String name) =>
        (headers[name.toLowerCase()] = request.headers[name]!));
    headers['host'] = [host]; // Host is a builtin header
    List<String> headerNames = headers.keys.toList()..sort();
    String canonicalHeaders = headerNames
        .map((s) =>
            (headers[s]!.map((v) => ('${s}:${_trimAll(v)}')).join('\n') + '\n'))
        .join();

    String signedHeaders = headerNames.join(';');

    // Build canonical query string
    Map<String, String> queryParameters = Map<String, String>()
      ..addAll(request.uri.queryParameters);
    List<String> queryKeys = queryParameters.keys.toList()..sort();
    String canonicalQueryString = queryKeys
        .map((s) => '${_uriEncode(s)}=${_uriEncode(queryParameters[s]!)}')
        .join('&');

    // Sign headers
    String canonicalRequest =
        '${httpMethod}\n${canonicalURI}\n${canonicalQueryString}\n${canonicalHeaders}\n${signedHeaders}\n$hashedPayloadStr';
    // print('\n>>>>>> canonical request \n' + canonicalRequest + '\n<<<<<<\n');

    Digest canonicalRequestHash = sha256.convert(utf8.encode(
        canonicalRequest)); //_hmacSha256.convert(utf8.encode(canonicalRequest));

    String stringToSign =
        'AWS4-HMAC-SHA256\n${dateIso8601}\n${dateYYYYMMDD}/${region}/${service}/aws4_request\n$canonicalRequestHash';
    // print('\n>>>>>> string to sign \n' + stringToSign + '\n<<<<<<\n');

    Digest dateKey = Hmac(sha256, utf8.encode("AWS4${secretKey}"))
        .convert(utf8.encode(dateYYYYMMDD));
    Digest dateRegionKey =
        Hmac(sha256, dateKey.bytes).convert(utf8.encode(region));
    Digest dateRegionServiceKey =
        Hmac(sha256, dateRegionKey.bytes).convert(utf8.encode(service));
    Digest signingKey = Hmac(sha256, dateRegionServiceKey.bytes)
        .convert(utf8.encode("aws4_request"));

    Digest signature =
        Hmac(sha256, signingKey.bytes).convert(utf8.encode(stringToSign));

    // Set signature in header
    request.headers.add('Authorization',
        'AWS4-HMAC-SHA256 Credential=${credential}, SignedHeaders=${signedHeaders}, Signature=$signature');
  }

  @protected
  Map<String, String> composeChunkRequestHeaders(
      {required Uri uri,
      required String dateIso8601,
      required String dateYYYYMMDD,
      required String contentType,
      required int contentLength,
      required int chunkContentLengthWithMeta,
      Permissions? permissions,
      Map<String, String>? meta}) {
    // Build canonical request
    String httpMethod = 'PUT';
    String canonicalURI = uri.path;
    String host = uri.host;

    String credential =
        '${accessKey}/${dateYYYYMMDD}/${region}/${service}/aws4_request';
    Map<String, String> resultHeaders = {
      'x-amz-date': dateIso8601, // Set date in header
      'x-amz-content-sha256':
          'STREAMING-AWS4-HMAC-SHA256-PAYLOAD', // Set payload hash in header
      'Content-Encoding': 'aws-chunked',
      'Content-Type': contentType,
      // 'Expect': '100-continue',
      // 'x-amz-storage-class': 'REDUCED_REDUNDANCY',
      'Content-Length': chunkContentLengthWithMeta.toString(),
      'x-amz-decoded-content-length': contentLength.toString()
    };

    if (meta != null) {
      for (MapEntry<String, String> me in meta.entries) {
        resultHeaders["x-amz-meta-${me.key}"] = me.value;
      }
    }
    if (permissions == Permissions.public) {
      resultHeaders['x-amz-acl'] = 'public-read';
    }
    // Build canonical headers string
    Map<String, List<String>> headers = Map<String, List<String>>();
    resultHeaders.keys.forEach((String name) => (headers[name.toLowerCase()] =
        resultHeaders[name] is List
            ? resultHeaders[name] as List<String>
            : [resultHeaders[name]!]));
    headers['host'] = [host]; // Host is a builtin header
    List<String> headerNames = headers.keys.toList()..sort();
    String canonicalHeaders = headerNames
        .map((s) =>
            (headers[s]!.map((v) => ('${s}:${_trimAll(v)}')).join('\n') + '\n'))
        .join();

    String signedHeaders = headerNames.join(';');

    // Build canonical query string
    Map<String, String> queryParameters = Map<String, String>()
      ..addAll(uri.queryParameters);
    List<String> queryKeys = queryParameters.keys.toList()..sort();
    String canonicalQueryString = queryKeys
        .map((s) => '${_uriEncode(s)}=${_uriEncode(queryParameters[s]!)}')
        .join('&');

    // Sign headers
    String canonicalRequest =
        '${httpMethod}\n${canonicalURI}\n${canonicalQueryString}\n${canonicalHeaders}\n${signedHeaders}\nSTREAMING-AWS4-HMAC-SHA256-PAYLOAD';
    // print('\n>>>>>> canonical request \n' + canonicalRequest + '\n<<<<<<\n');

    Digest canonicalRequestHash = sha256.convert(utf8.encode(
        canonicalRequest)); //_hmacSha256.convert(utf8.encode(canonicalRequest));

    String stringToSign =
        'AWS4-HMAC-SHA256\n${dateIso8601}\n${dateYYYYMMDD}/${region}/${service}/aws4_request\n$canonicalRequestHash';
    // print('\n>>>>>> string to sign \n' + stringToSign + '\n<<<<<<\n');

    Digest dateKey = Hmac(sha256, utf8.encode("AWS4${secretKey}"))
        .convert(utf8.encode(dateYYYYMMDD));
    Digest dateRegionKey =
        Hmac(sha256, dateKey.bytes).convert(utf8.encode(region));
    Digest dateRegionServiceKey =
        Hmac(sha256, dateRegionKey.bytes).convert(utf8.encode(service));
    Digest signingKey = Hmac(sha256, dateRegionServiceKey.bytes)
        .convert(utf8.encode("aws4_request"));

    Digest signature =
        Hmac(sha256, signingKey.bytes).convert(utf8.encode(stringToSign));

    // Set signature in header
    resultHeaders['Authorization'] =
        'AWS4-HMAC-SHA256 Credential=${credential}, SignedHeaders=${signedHeaders}, Signature=$signature';

    return resultHeaders;
  }

  String calculateChunkedSignature(
    List<int> data,
    String prevSignature, {
    required String dateIso8601,
    required String dateYYYYMMDD,
  }) {
    String stringToSign =
        'AWS4-HMAC-SHA256-PAYLOAD\n${dateIso8601}\n${dateYYYYMMDD}/${region}/${service}/aws4_request\n$prevSignature\n${sha256.convert([])}\n${sha256.convert(data)}';
    // print('\n>>>>>> chunk string to sign \n' + stringToSign + '\n<<<<<<\n');

    Digest dateKey = Hmac(sha256, utf8.encode("AWS4${secretKey}"))
        .convert(utf8.encode(dateYYYYMMDD));
    Digest dateRegionKey =
        Hmac(sha256, dateKey.bytes).convert(utf8.encode(region));
    Digest dateRegionServiceKey =
        Hmac(sha256, dateRegionKey.bytes).convert(utf8.encode(service));
    Digest signingKey = Hmac(sha256, dateRegionServiceKey.bytes)
        .convert(utf8.encode("aws4_request"));

    Digest signature =
        Hmac(sha256, signingKey.bytes).convert(utf8.encode(stringToSign));

    return '$signature';
  }
}


lib/src/spaces.dart
import 'dart:async';
import 'package:http_client/console.dart' as http;
import 'package:xml/xml.dart' as xml;

import 'client.dart';
import 'bucket.dart';

enum Provider {
  amazon,
  yandex,
}

class Spaces extends Client {
  final Provider provider;
  late String _endpointUrl;
  Spaces(
      {required String region,
      required String accessKey,
      required String secretKey,
      String? sessionToken,
      http.Client? httpClient,
      this.provider = Provider.amazon})
      : super(
            region: region,
            accessKey: accessKey,
            secretKey: secretKey,
            sessionToken: sessionToken,
            service: "s3",
            httpClient: httpClient);

  Bucket bucket(String bucket) {
    switch (provider) {
      case Provider.amazon:
        _endpointUrl = "https://s3.${region}.amazonaws.com";
        break;
      case Provider.yandex:
        _endpointUrl = "https://storage.yandexcloud.net";
        break;
      default:
        throw Exception(
            "Endpoint URL not supported. Create Bucket client manually.");
    }
    return Bucket(
        region: region,
        accessKey: accessKey,
        secretKey: secretKey,
        endpointUrl: "$_endpointUrl/${bucket}",
        sessionToken: sessionToken,
        httpClient: httpClient);
  }

  Future<List<String>> listAllBuckets() async {
    xml.XmlDocument doc = await getUri(Uri.parse(_endpointUrl + '/'));
    List<String> res = <String>[];
    for (xml.XmlElement root in doc.findElements('ListAllMyBucketsResult')) {
      for (xml.XmlElement buckets in root.findElements('Buckets')) {
        for (xml.XmlElement bucket in buckets.findElements('Bucket')) {
          for (xml.XmlElement name in bucket.findElements('Name')) {
            res.add(name.text);
          }
        }
      }
    }
    return res;
  }
}


lib/src/stream_transformers.dart
import 'dart:async';

import 'package:aws_s3_client/src/client.dart';

/// StreamTransformer which collects values until it will exceed [chunkSize].
/// Emits chunks of [chunkSize] except the last one, which can be smaller
class ChunkTransformer implements StreamTransformer<List<int>, List<int>> {
  late int _chunkSize;
  late StreamController _controller;
  // ignore: cancel_subscriptions
  late StreamSubscription? _subscription;
  bool? cancelOnError;
  late Stream<List<int>> _stream;
  final _buffer = <int>[];

  ChunkTransformer(
      {int chunkSize = defaultChunkSize,
      bool sync = false,
      this.cancelOnError}) {
    _chunkSize = chunkSize;
    _controller = StreamController<List<int>>(
        onListen: _onListen,
        onCancel: _onCancel,
        onPause: () {
          _subscription!.pause();
        },
        onResume: () {
          _subscription!.resume();
        },
        sync: sync);
  }

  ChunkTransformer.broadcast(
      {int chunkSize = defaultChunkSize,
      bool sync = false,
      this.cancelOnError}) {
    _chunkSize = chunkSize;
    _controller = StreamController<List<int>>.broadcast(
        onListen: _onListen, onCancel: _onCancel, sync: sync);
  }

  void _onListen() {
    _subscription = _stream.listen(onData,
        onError: _controller.addError,
        onDone: _onDone,
        cancelOnError: cancelOnError);
  }

  void _onCancel() {
    _subscription!.cancel();
    _subscription = null;
  }

  void onData(List<int> data) {
    _buffer.addAll(data);
    _emitData();
  }

  void _onDone() {
    if (_buffer.isNotEmpty) {
      _controller.add(_buffer);
    }
    _subscription!.cancel();
    _subscription = null;
    _controller.close();
  }

  void _emitData() {
    if (_buffer.length >= _chunkSize) {
      _controller.add(_buffer.sublist(0, _chunkSize));
      _buffer.removeRange(0, _chunkSize);
      _emitData();
    }
  }

  Stream<List<int>> bind(Stream<List<int>> stream) {
    this._stream = stream;
    return _controller.stream as Stream<List<int>>;
  }

  StreamTransformer<RS, RT> cast<RS, RT>() => StreamTransformer.castFrom(this);
}


lib/src/bucket.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:http_client/console.dart' as http_client;
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart' as xml;

import 'client.dart';
import 'results.dart';
import 'stream_transformers.dart';

enum Permissions {
  private,
  public,
}

class Bucket extends Client {
  final int chunkSize;
  final String endpointUrl;
  Bucket(
      {required String region,
      required String accessKey,
      required String secretKey,
      required this.endpointUrl,
      String? sessionToken,
      this.chunkSize = defaultChunkSize,
      http_client.Client? httpClient})
      : super(
            region: region,
            accessKey: accessKey,
            secretKey: secretKey,
            service: "s3",
            sessionToken: sessionToken,
            httpClient: httpClient);

  /// List the Bucket's Content
  Stream<BucketContent> listContent(
      {String? delimiter, String? prefix, int? maxKeys}) async* {
    late bool isTruncated;
    String? marker;
    do {
      Uri uri = Uri.parse(endpointUrl + '/');
      Map<String, dynamic> params = Map<String, dynamic>();
      if (delimiter != null) params['delimiter'] = delimiter;
      if (marker != null) {
        params['marker'] = marker;
        marker = null;
      }
      if (maxKeys != null) params['max-keys'] = "$maxKeys";
      if (prefix != null) params['prefix'] = prefix;
      uri = uri.replace(queryParameters: params);
      xml.XmlDocument doc = await getUri(uri);
      String? lastKey;
      for (xml.XmlElement root in doc.findElements('ListBucketResult')) {
        for (xml.XmlNode node in root.children) {
          if (node is xml.XmlElement) {
            xml.XmlElement element = node;
            switch ('${element.name}') {
              case "NextMarker":
                marker = element.text;
                break;
              case "IsTruncated":
                isTruncated = element.text.toLowerCase() != "false" &&
                    element.text != "0";
                break;
              case "Contents":
                late String key;
                late DateTime lastModifiedUtc;
                late String eTag;
                late int size;
                for (xml.XmlNode node in element.children) {
                  if (node is xml.XmlElement) {
                    xml.XmlElement element = node;
                    switch ('${element.name}') {
                      case "Key":
                        key = element.text;
                        lastKey = key;
                        break;
                      case "LastModified":
                        lastModifiedUtc = DateTime.parse(element.text);
                        break;
                      case "ETag":
                        eTag = element.text;
                        break;
                      case "Size":
                        size = int.parse(element.text);
                        break;
                    }
                  }
                }
                yield BucketContent(
                  key: key,
                  lastModifiedUtc: lastModifiedUtc,
                  eTag: eTag,
                  size: size,
                );
                break;
            }
          }
        }
      }
      if (isTruncated && lastKey != null) marker = lastKey;
    } while (isTruncated);
  }

  /// Uploads file stream. Returns Etag.
  Future<String> uploadFileStream(String key, Stream<List<int>> fileStream,
      int contentLength, String contentType, Permissions permissions,
      {Map<String, String>? meta}) async {
    bool isFirstChunk = true;
    String signature;
    String? prevSignature;
    Uri uri = Uri.parse(endpointUrl + '/' + key);

    DateTime date = DateTime.now().toUtc();

    // String dateIso8601 = "20130524T000000Z";
    String dateIso8601 = date.toIso8601String();
    dateIso8601 = dateIso8601
            .substring(0, dateIso8601.indexOf('.'))
            .replaceAll(':', '')
            .replaceAll('-', '') +
        'Z';

    // String dateYYYYMMDD = "20130524";
    String dateYYYYMMDD = date.year.toString().padLeft(4, '0') +
        date.month.toString().padLeft(2, '0') +
        date.day.toString().padLeft(2, '0');

    int contentLengthWithMeta =
        calculateContentLengthWithMeta(contentLength, chunkSize);

    final headers = composeChunkRequestHeaders(
        uri: uri,
        dateYYYYMMDD: dateYYYYMMDD,
        dateIso8601: dateIso8601,
        contentLength: contentLength,
        contentType: contentType,
        permissions: permissions,
        chunkContentLengthWithMeta: contentLengthWithMeta,
        meta: meta);
    String canonicalRequestSignature =
        headers['Authorization']!.split('Signature=').last;
    prevSignature = null;

    http.StreamedRequest request = http.StreamedRequest('PUT', uri);
    request.headers.addAll(headers);
    final futureRequest = request.send();

    Future<String> sendChunkRequest(List<int> data) async {
      signature = calculateChunkedSignature(
        data,
        isFirstChunk ? canonicalRequestSignature : prevSignature!,
        dateYYYYMMDD: dateYYYYMMDD,
        dateIso8601: dateIso8601,
      );
      if (isFirstChunk) {
        isFirstChunk = false;
      }
      request.sink.add(data.length.toRadixString(16).toString().codeUnits);
      request.sink.add(";chunk-signature=$signature\r\n".codeUnits);
      if (data.isNotEmpty) request.sink.add(data);
      request.sink.add("\r\n".codeUnits);
      if (data.isEmpty) {
        request.sink.close();
        final responseStream = await futureRequest;
        final response = await http.Response.fromStream(responseStream);
        // print(response.statusCode);
        // print(response.reasonPhrase);
        // print(response.headers);
        // print(await response.transform(utf8.decoder).first);
        return response.headers[HttpHeaders.etagHeader]!;
      }
      prevSignature = signature;
      return '';
    }

    Future<String> sendChunkRequestSync(List<int> val,
        [Future? previousChunk]) async {
      if (previousChunk != null) {
        await previousChunk;
      }
      return sendChunkRequest(val);
    }

    Future<dynamic> handleFileStream(Stream<List<int>> fileStream) {
      late Future prevChunk;

      final completer = Completer();
      fileStream.listen((val) {
        prevChunk = sendChunkRequestSync(val, prevChunk);
      }, onDone: () {
        sendChunkRequestSync([], prevChunk).then((etag) {
          completer.complete(etag);
        });
      });
      return completer.future;
    }

    return await (handleFileStream(
            fileStream.transform(ChunkTransformer(chunkSize: chunkSize)))
        as FutureOr<String>);
  }

  int calculateContentLengthWithMeta(int contentLength, int chunkSize) =>
      (chunkSize.toRadixString(16).codeUnits.length + 85) *
          (contentLength / chunkSize).floor() +
      (contentLength % chunkSize).toRadixString(16).codeUnits.length +
      85 +
      86 +
      contentLength;

  /// Upload file. Return Etag.
  Future<String> uploadFile(String key, Uint8List content, String contentType,
      Permissions permissions,
      {Map<String, String>? meta}) async {
    int contentLength = content.lengthInBytes;

    Digest contentSha256 = sha256.convert(content);

    String uriStr = endpointUrl + '/' + key;
    http_client.Request request = http_client.Request('PUT', Uri.parse(uriStr),
        headers: http_client.Headers(), body: content);

    if (meta != null) {
      for (MapEntry<String, String> me in meta.entries) {
        request.headers.add("x-amz-meta-${me.key}", me.value);
      }
    }

    if (sessionToken != null) {
      request.headers.add('x-amz-security-token', sessionToken);
    }

    if (permissions == Permissions.public) {
      request.headers.add('x-amz-acl', 'public-read');
    }
    request.headers.add('Content-Length', contentLength);
    request.headers.add('Content-Type', contentType);
    signRequest(request, contentSha256: contentSha256);
    http_client.Response response = await httpClient.send(request);
    BytesBuilder builder = BytesBuilder(copy: false);
    await response.body.forEach(builder.add);
    String body = utf8.decode(builder.toBytes()); // Should be empty when OK
    if (response.statusCode != 200) {
      throw ClientException(response.statusCode, response.reasonPhrase,
          response.headers.toSimpleMap(), body);
    }
    return response.headers[HttpHeaders.etagHeader]!.first;
  }

  /// Delete file
  Future<void> delete(String key) async {
    String uriStr = endpointUrl + '/' + key;
    http_client.Request request = http_client.Request(
        'DELETE', Uri.parse(uriStr),
        headers: http_client.Headers());

    signRequest(request, contentSha256: sha256.convert([]));
    http_client.Response response = await httpClient.send(request);
    BytesBuilder builder = BytesBuilder(copy: false);
    await response.body.forEach(builder.add);
    String body = utf8.decode(builder.toBytes()); // Should be empty when OK
    if (response.statusCode != 200 && response.statusCode != 204) {
      throw ClientException(response.statusCode, response.reasonPhrase,
          response.headers.toSimpleMap(), body);
    }
  }
}
